src/lib.rs

```diff
@@ -0,0 +1,206 @@
+#![deny(warnings)]
+extern crate image;
+
+mod convolution;
+
+pub use convolution::{Convolution, ConvolutionError};
+
+use image::{DynamicImage, GrayImage, ImageBuffer};
+
+pub enum EdgeMode {
+    Extend,
+    Wrap,
+    Mirror,
+}
+
+pub fn convolve(
+    image_in: DynamicImage,
+    convolution: Convolution,
+    edge_mode: EdgeMode,
+) -> Result<GrayImage, ConvolutionError> {
+    let gray_image = image_in.to_luma();
+    let mut build = Vec::with_capacity((gray_image.width() * gray_image.height()) as usize);
```

Comment:
The `image` crate should implement `ExactSizeIterator` for `EnumeratePixels` and `Pixels` which would let you do `gray_image.pixels().len()`... Might be worth a PR to them.